File: CmacLib/src/Prediction.cpp
#include "Prediction.h"
#include "Result.h"

using namespace CmacLib;

void Prediction::SetValues(std::vector<double>& values)
{
	this->values = values;
}

void Prediction::SetActiveWeightIndices(std::vector<size_t>& activeWeightIndices)
{
	this->activeWeightIndices = activeWeightIndices;
}

void Prediction::SetActiveWeights(std::vector<std::vector<double>>& weights)
{
	this->activeWeights = weights;
}

void Prediction::SetBasisValues(std::vector<double>& basis)
{
	this->basisValues = basis;
}

void Prediction::SetResult(IResult* result)
{
	this->result.reset(result);
}

Prediction::Prediction()
	: values(std::vector<double>())
	, activeWeightIndices(std::vector<size_t>())
	, activeWeights(std::vector<std::vector<double>>())
	, basisValues(std::vector<double>())
	, result(std::make_unique<Result>()){}

Prediction::~Prediction()
{
	this->result.reset();
}

const std::vector<double>& Prediction::GetValues()
{
	return this->values;
}

const std::vector<size_t>& Prediction::GetActiveWeightIndices()
{
	return this->activeWeightIndices;
}

const std::vector<std::vector<double>>& Prediction::GetActiveWeights()
{
	return this->activeWeights;
}

const std::vector<double>& Prediction::GetBasisValues()
{
	return this->basisValues;
}

const IResult& Prediction::GetResult()
{
	return *(this->result);
}

File: langs/matlab/+CmacLib/+Implementations/Prediction.m
classdef (Hidden,Sealed) Prediction < CmacLib.Interfaces.IPrediction & CmacLib.Implementations.Result
    %PREDICTION Summary of this class goes here
    %   Detailed explanation goes here

    %% Constructor
    methods
        function obj = Prediction()
            %PREDICTION Construct an instance of this class
            %   Detailed explanation goes here
            obj@CmacLib.Implementations.Result();
            obj@CmacLib.Interfaces.IPrediction();
        end
    end

    %% Setters
    methods
        function SetValues(obj, values)
            arguments
                obj Prediction
                values (:,1) double
            end
            obj.Values = values;
        end

        function SetActiveWeightIndices(obj, indices)
            arguments
                obj Prediction
                indices (:,1) uint32
            end
            obj.ActiveWeightIndices = indices;
        end

        function SetActiveWeights(obj, weights)
            arguments
                obj Prediction
                weights (:,:) double
            end
            obj.ActiveWeights = weights;
        end

        function SetBasisValues(obj, basis)
            arguments
                obj Prediction
                basis (:,1) double
            end
            obj.BasisValues = basis;
        end

    end
    
    %% Converters
    methods
        function objIResult = IResult(obj)
            objIResult = CmacLib.Interfaces.IResult();
            objIResult.IsSuccessful = obj.IsSuccessful;
            objIResult.Message = obj.Message;
        end

        function objIPrediction = IPrediction(obj)
            objIPrediction = CmacLib.Interfaces.IPrediction();
            objIPrediction.IsSuccessful = obj.IsSuccessful;
            objIPrediction.Message = obj.Message;
            objIPrediction.Values = obj.Values;
            objIPrediction.ActiveWeightIndices = obj.ActiveWeightIndices;
            objIPrediction.ActiveWeights = obj.ActiveWeights;
            objIPrediction.BasisValues = obj.BasisValues;
        end
    end
end

